### DISCUSSION TITLE
Close/hide a window

### DISCUSSION THREAD

**Discussion by russoue** [slint-ui/slint#536] [Q&A]
Sorry for this noob question.  I am trying to play with sixtyfps and had the following simple dialog in my code:

```rust
sixtyfps::sixtyfps! {
    import { StandardButton } from "sixtyfps_widgets.60";

    ConfigWindow := Window {
        title: "Hello world";
        width: 320px;
        height: 64px;
        callback user_answered(bool);
        VerticalLayout {
            Text {
                text: "Do you want to create a default configuration?";
                color: black;
            }
            HorizontalLayout {
                StandardButton {
                    kind: yes;
                    clicked => { self.text = "Yes!!!"; root.user_answered(true); }
                }
                StandardButton {
                    kind: no;
                    clicked => { self.text = "No!!!"; root.user_answered(false); }
                }
            }
        }
    }
}
```

I have the corresponding Rust code:

```rust
    let user_answer = Arc::new(Mutex::new(false));
    let config_window = ConfigWindow::new();
    let user_answer_clone = Arc::clone(&user_answer);
    config_window.on_user_answered(move |answer| {
        *user_answer_clone.lock().unwrap() = answer;
    });
    config_window.run();
```

This works but I want to close the window after the user clicks any of those buttons.  So far I could not find a way that works.  The closest I got is calling `hide()` in the callback:

```rust
    ...
    let config_window_weak = config_window.as_weak();
    config_window.on_user_answered(move |answer| {
        *user_answer_clone.lock().unwrap() = answer;
        let config_window = config_window_weak.unwrap();
        config_window.hide();
    });
    ...
```

But the above encounters a runtime borrow checker error: `thread 'main' panicked at 'already borrowed: BorrowMutError'`.

What's the best way to close a window from a callback?  Also, is there a dialog widget already which does what I am trying to do: ask a question and collect a yes/no or ok/cancel answer from the user?

**Comment by tronical** [comment]
That’s a perfectly fine question, not noob at all :). In fact, I think calling hide() is fine, it’s a bug that you’re getting a runtime error. I’ll take a look at that and see if I can reproduce it.

**Reply by tronical** [reply]
Update: The fix for this landed in the master branch. Let us know if this works for you :)

**Reply by russoue** [reply]
Thanks for the quick fix!  The window hides fine now.  Is there any way to close the window altogether?

**Reply by tronical** [reply]
The window belongs to the component, so when the last reference to the component vanishes, the window is also destroyed.

Or are you perhaps looking for [sixtyfps::quit_event_loop()](https://sixtyfps.io/releases/0.1.2/docs/rust/sixtyfps/fn.quit_event_loop.html)?

**Reply by russoue** [reply]
That also works, thank you.
